pkg/api/config.go

```diff
@@ -3,71 +3,230 @@ package api
-				typeCount += 1
-				errs = append(errs, validateTargetCloud(test.As, testConfig.TargetCloud))
-			}
-			if testConfig := test.OpenshiftInstallerClusterTestConfiguration; testConfig != nil {
-				typeCount += 1
-				errs = append(errs, validateTargetCloud(test.As, testConfig.TargetCloud))
-			}
-			if testConfig := test.OpenshiftInstallerSmokeClusterTestConfiguration; testConfig != nil {
-				typeCount += 1
-				errs = append(errs, validateTargetCloud(test.As, testConfig.TargetCloud))
-			}
-			if typeCount == 0 {
-				// TODO remove when the migration is completed
-				if len(test.From) == 0 {
-					errs = append(errs, fmt.Errorf("test %q has no type", test.As))
-				}
-			} else if typeCount == 1 {
-				// TODO remove when the migration is completed
-				if len(test.From) > 0 {
-					errs = append(errs, fmt.Errorf("test %q specifies both `From` and a test type", test.As))
-				}
-			} else if typeCount > 1 {
-				errs = append(errs, fmt.Errorf("test %q has more than one type", test.As))
-			}
+	var validationErrors []error
+
+	validationErrors = append(validationErrors, validateReleaseBuildConfiguration(config))
+	validationErrors = append(validationErrors, validateBuildRootImageConfiguration("build_root", config.InputConfiguration.BuildRootImage))
+	validationErrors = append(validationErrors, validateTestStepConfiguration("tests", config.Tests))
+
+	if config.InputConfiguration.BaseImages != nil {
+		validationErrors = append(validationErrors, validateImageStreamTagReferenceMap("base_images", config.InputConfiguration.BaseImages))
+	}
+
+	if config.InputConfiguration.BaseRPMImages != nil {
+		validationErrors = append(validationErrors, validateImageStreamTagReferenceMap("base_rpm_images", config.InputConfiguration.BaseRPMImages))
+	}
+
+	if config.PromotionConfiguration != nil && config.InputConfiguration.ReleaseTagConfiguration != nil {
+		validationErrors = append(validationErrors, validatePromotionWithTagSpec(*config.PromotionConfiguration, *config.InputConfiguration.ReleaseTagConfiguration))
+	} else if config.PromotionConfiguration == nil && config.InputConfiguration.ReleaseTagConfiguration != nil {
+		validationErrors = append(validationErrors, validateReleaseTagConfiguration("tag_specification", *config.InputConfiguration.ReleaseTagConfiguration))
+	} else if config.PromotionConfiguration != nil && config.InputConfiguration.ReleaseTagConfiguration == nil {
+		validationErrors = append(validationErrors, validatePromotionConfiguration("promotion", *config.PromotionConfiguration))
+	}
+
+	return kerrors.NewAggregate(validationErrors)
+}
+
+func validatePromotionWithTagSpec(promotion PromotionConfiguration, tagSpec ReleaseTagConfiguration) error {
+	var validationErrors []error
+
+	if len(promotion.Name) == 0 && len(promotion.Tag) == 0 {
+		if err := validateReleaseTagConfiguration("tag_specification", tagSpec); err != nil {
+			validationErrors = append(validationErrors, fmt.Errorf("promotion: %v", err))
```

Comment:
We should just write out the checks -- because today I think the error a user is going to get is confusing.


```go
func validatePromotionWithTagSpec(promotion PromotionConfiguration, tagSpec ReleaseTagConfiguration) error {
	var validationErrors []error

	if len(promotion.Name) == 0 && len(promotion.Tag) == 0 {
		if len(tagSpec.Name) != 0 || len(tagSpec.Tag) != 0 {
			// will get defaulted, is ok
		} else {
			validationErrors = append(validationErrors, errors.New("promotion: no name or tag provided and could not derive defaults from tag_specification"))
		}
	} else {
		validationErrors = append(validationErrors, validatePromotionConfiguration("promotion", promotion))
	}

	return kerrors.NewAggregate(validationErrors)
}
```